### PR TITLE
[Runtime Epoch Split] (4/n) Make ShardTracker accessible from RuntimeWithEpochManagerAdapter.

### PR DESCRIPTION
**Pull request by robin-near**
As a reminder of the overall goal, we want to split `RuntimeWithEpochManagerAdapter`, so that any code that needs the runtime will use an `Arc<RuntimeAdapter>`, and any code that uses the epoch manager will use the `EpochManagerHandle`.

We're doing this refactoring bottom-up, i.e. propagating `RuntimeWithEpochManagerAdapter` around at the top-level, but making some lower-level components use `Arc<RuntimeAdapter>` and/or `EpochManagerHandle`.

Now, it turns out that `ShardsManager`, one of the lower-level components, would need not only `EpochManagerHandle`, but also an additional `ShardTracker` built on top of `EpochManagerHandle`, whose functionality was previously covered by `RuntimeWithEpochManagerAdapter` interface. `ShardTracker` is not technically part of `EpochManager` either, because it isn't inherent to the epoch part of the protocol, but is controlled by the node operator's config options.

`ShardTracker` is a field of `NightshadeRuntime`, but is not available in the `KeyValueRuntime`; instead, the `KeyValueRuntime` implements all the `ShardTracker` functionalities via the `RuntimeWithEpochManagerAdapter` interface. This would not work.

This PR moves `ShardTracker` to the near-epoch-manager crate, adds an additional `shard_tracker()` function to `RuntimeWithEpochManagerAdapter` to allow passing it down from a higher-level component to a lower-level component, and performs a refactoring to make `KeyValueRuntime` provide the `ShardTracker` functionality via its `EpochManagerAdapter` implementation. This also required changing `ShardTracker` to use `EpochManagerAdapter`, as opposed to the concrete `EpochManagerHandle`.

### CODE REVIEW

**Comment by akhi3030** [comment]
@robin-near: do you know if someone from core team will review this or do you want me to find a reviewer for you?

**Comment by robin-near** [comment]
> @robin-near: do you know if someone from core team will review this or do you want me to find a reviewer for you?

Hmm, I was expecting the core-team alias to resolve to an engineer. Let me try again.

File: chain/epoch-manager/src/shard_tracker.rs [hunk 1]
```diff
@@ -31,17 +34,21 @@ type BitMask = Vec<bool>;
...
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
 }
 
 impl ShardTracker {
-    pub fn new(tracked_config: TrackedConfig, epoch_manager: EpochManagerHandle) -> Self {
-        ShardTracker { tracked_config, tracking_shards: AppendOnlyMap::new(), epoch_manager }
+    pub fn new(tracked_config: TrackedConfig, epoch_manager: Arc<dyn EpochManagerAdapter>) -> Self {
+        ShardTracker {
+            tracked_config,
+            tracking_shards_cache: Arc::new(SyncLruCache::new(1024)),
```
[the remark this answers is not in the document]
Comment (robin-near) [reply chain/epoch-manager/src/shard_tracker.rs:52]: I missed that - thanks!
